File: day4/typescript/part_2.ts
import * as fs from "fs";

function countXmasXShape(grid: string[][]): number {
  let count = 0;
  const rows = grid.length;
  const cols = grid[0].length;

  function checkXmasXShape(x: number, y: number): boolean {
    if (grid[x][y] !== "A") {
      return false;
    }

    const topLeft = grid[x - 1][y - 1] + grid[x + 1][y + 1];
    const topRight = grid[x - 1][y + 1] + grid[x + 1][y - 1];

    return (
      (topLeft === "MS" || topLeft === "SM") &&
      (topRight === "MS" || topRight === "SM")
    );
  }

  for (let x = 1; x < rows - 1; x++) {
    for (let y = 1; y < cols - 1; y++) {
      if (checkXmasXShape(x, y)) {
        count++;
      }
    }
  }

  return count;
}

const gridXShape: string[][] = fs
  .readFileSync("input.txt", "utf-8")
  .split("\n")
  .map((line) => line.split(""));

console.log(countXmasXShape(gridXShape));

File: day4/go/part_2.go
package main

import (
	"bufio"
	"fmt"
	"os"
)

func countXmas(grid [][]rune) int {
	ret := 0

	checkXmasXShape := func(x, y int) bool {
		if grid[x][y] != 'A' {
			return false
		}

		topLeft := string([]rune{grid[x-1][y-1], grid[x+1][y+1]})
		topRight := string([]rune{grid[x-1][y+1], grid[x+1][y-1]})

		return (topLeft == "MS" || topLeft == "SM") &&
			(topRight == "MS" || topRight == "SM")
	}

	for x := 1; x < len(grid)-1; x++ {
		for y := 1; y < len(grid[0])-1; y++ {
			if checkXmasXShape(x, y) {
				ret++
			}
		}
	}

	return ret
}

func main() {
	file, err := os.Open("input.txt")
	if err != nil {
		panic(err)
	}
	defer file.Close()

	var grid [][]rune
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		grid = append(grid, []rune(scanner.Text()))
	}

	fmt.Println(countXmas(grid))
}
